nanocurrency/nano-node
# Improve: Consolidate artifact build process for all environments & Add Workflow for preparing release branch

### What will the build pipeline look like ?
- Twice per week the new “Build all artifacts” workflow runs automatically and creates a new V{Major}.0DBX build for Test, Beta and Live, where X is incremented by 1 if a new commit exists compared to the previous tag.
- We could also run the “Build all artifacts” workflow manually 
  - to create a new DB build in case we don’t want to wait for the automated execution
  - recreate a previous DB build (by setting increment to 0 for example)

### What will the release process look like ?
- We run the “Prepare release” workflow on the tag (or branch) which we want to use as starting point for the newly created release branch.
  - Usually this would be a V{Major}.0DBX tag that creates a releases/v{Major} branch

- After the  release branch has been successfully created we can run the “Build all artifacts” workflow manually to generate the release builds.
  - On its first execution it will create a tag V{Major}.0  (the increment which is 1 by default is ignored, to make sure that we have the correct tag even if we forget to set increment 0)
  - We can execute the build process again by setting increment=0 in case the build failed for whatever reason
  - We can create the next minor version V{Major}.1 by leaving the default (increment=1)


### Improvements :
- repository and branch agnostic
- single workflow to build artifacts for all environments
- new workflow to prepare the next releases branch from an existing branch or tag

### Possible simplification
We could get rid of the “increment” input in the workflow by adopting a similar behaviour as DB builds.
When “Build all artifacts” runs on a release branch,
- if there  is no new commit, the workflow builds binaries & docker images for the current tag
- If there is a new commit compared to the previous tag, the workflow increments version_minor by 1, creates a new tag and builds the new binaries & docker images


### Prerequisits before merge
Required secrets:
```
AWS_ACCESS_KEY_ID 
AWS_SECRET_ACCESS_KEY
```

Optional secrets
```
DOCKER_PASSWORD (required to upload images to hub.docker.com)
```

Optional Repository variables:
```
DOCKER_REGISTRY (default = "nanocurrency")
DOCKER_USER  (default = "nanoreleaseteam")
S3_BUCKET_NAME (default = "repo.nano.org")
S3_BUILD_DIRECTORY (default = "")
```


### How was the PR tested ?

- Run "All (Live Beta Test)" manually on branch prs/unified_artifacts_worflow
https://github.com/gr0vity-dev/nano-node/actions/runs/5444908141/jobs/9903410239

- run "Prepare Release" from {tag} to create releases/v26
https://github.com/gr0vity-dev/nano-node/actions/runs/5444720099/jobs/9902956258

- run "Prepare Release" again. It's expected to fail because the releases/v26 already exists.
https://github.com/gr0vity-dev/nano-node/actions/runs/5445139689/jobs/9903953960

- Run "All (Live Beta Test)" manually on tag V26.0
https://github.com/gr0vity-dev/nano-node/actions/runs/5446028132
Creates docker images:
```
docker pull gr0vity/nano-test:V26.0
docker pull gr0vity/nano-beta:V26.0
docker pull gr0vity/nano:V26.0
```

Creates binaries:
```
Test:
https://gr0vity-nanonode.s3.us-east-2.amazonaws.com/test/binaries/nano-node-V26.0-win64.zip
https://gr0vity-nanonode.s3.us-east-2.amazonaws.com/test/binaries/nano-node-V26.0-win64.exe
https://gr0vity-nanonode.s3.us-east-2.amazonaws.com/test/binaries/nano-node-V26.0-Darwin.dmg
https://gr0vity-nanonode.s3.us-east-2.amazonaws.com/test/binaries/nano-node-V26.0-Linux.deb
https://gr0vity-nanonode.s3.us-east-2.amazonaws.com/test/binaries/nano-node-V26.0-Linux.tar.bz2


Beta:
https://gr0vity-nanonode.s3.us-east-2.amazonaws.com/beta/binaries/nano-node-V26.0-Linux.deb
https://gr0vity-nanonode.s3.us-east-2.amazonaws.com/beta/binaries/nano-node-V26.0-Darwin.dmg
https://gr0vity-nanonode.s3.us-east-2.amazonaws.com/beta/binaries/nano-node-V26.0-Linux.tar.bz2
https://gr0vity-nanonode.s3.us-east-2.amazonaws.com/beta/binaries/nano-node-V26.0-win64.exe
https://gr0vity-nanonode.s3.us-east-2.amazonaws.com/beta/binaries/nano-node-V26.0-win64.zip
https://gr0vity-nanonode.s3.us-east-2.amazonaws.com/beta/source/nanocurrency-beta-26.0-1.el8.src.rpm


Live:
https://gr0vity-nanonode.s3.us-east-2.amazonaws.com/live/binaries/nano-node-V26.0-Linux.deb
https://gr0vity-nanonode.s3.us-east-2.amazonaws.com/live/binaries/nano-node-V26.0-Darwin.dmg
https://gr0vity-nanonode.s3.us-east-2.amazonaws.com/live/binaries/nano-node-V26.0-Linux.tar.bz2
https://gr0vity-nanonode.s3.us-east-2.amazonaws.com/live/binaries/nano-node-V26.0-win64.exe
https://gr0vity-nanonode.s3.us-east-2.amazonaws.com/live/binaries/nano-node-V26.0-win64.zip
https://gr0vity-nanonode.s3.us-east-2.amazonaws.com/live/source/nanocurrency-beta-26.0-1.el8.src.rpm
```

### What needs to be tested ?
- cron execution on develop branch
- `increment`= 0 on a DB build
- `increment` with values other than 0 or 1

### Flowcharts of the intended behaviour for “Build all artifacts” workflow
![Consolidate_workflow_part1](https://github.com/nanocurrency/nano-node/assets/85646666/d4b443cc-055b-4f4f-84c3-dc46b03cfa6f)
![Consolidate_workflow_part2](https://github.com/nanocurrency/nano-node/assets/85646666/861d6894-b2e5-4db3-a2c6-3dd593365e01)

### Flowcharts of the intended behaviour for “Prepare release” workflow
![Prepare_release](https://github.com/nanocurrency/nano-node/assets/85646666/b1ce715e-eaf9-4687-a250-d150f796efe0)
